src/interface.jl

```diff
@@ -31,111 +46,201 @@ ADBackend(::Val) = error("The requested AD backend is not available. Make sure t
 this is not the case. For example the *inverse* evaluation of `PlanarLayer`
 requires an iterative procedure to evaluate.
 """
-isclosedform(b::Bijector) = true
+isclosedform(t::Transform) = true
 
+# Invertibility "trait".
 """
-    inverse(b::Bijector)
-    Inverse(b::Bijector)
+    Invertible
 
-A `Bijector` representing the inverse transform of `b`.
+Represents the trait of being, well, non-invertible.
+"""
+struct NotInvertible end
 """
-struct Inverse{B <: Bijector, N} <: Bijector{N}
-    orig::B
+    Invertible
 
-    Inverse(b::B) where {N, B<:Bijector{N}} = new{B, N}(b)
-end
+Represents the trait of being, well, invertible.
+"""
+struct Invertible end
 
-# field contains nested numerical parameters
-Functors.@functor Inverse
+# Useful for checking if compositions, etc. are invertible or not.
+Base.:+(::NotInvertible, ::Invertible) = NotInvertible()
+Base.:+(::Invertible, ::NotInvertible) = NotInvertible()
+Base.:+(::NotInvertible, ::NotInvertible) = NotInvertible()
+Base.:+(::Invertible, ::Invertible) = Invertible()
 
-up1(b::Inverse) = Inverse(up1(b.orig))
+"""
+    invertible(t)
 
-inverse(b::Bijector) = Inverse(b)
-inverse(ib::Inverse{<:Bijector}) = ib.orig
-Base.:(==)(b1::Inverse{<:Bijector}, b2::Inverse{<:Bijector}) = b1.orig == b2.orig
+Return `Invertible()` if `t` is invertible, and `NotInvertible()` otherwise.
+"""
+invertible(::Transform) = NotInvertible()
+invertible(f::Elementwise) = invertible(f.x)
 
 """
-    logabsdetjac(b::Bijector, x)
-    logabsdetjac(ib::Inverse{<:Bijector}, y)
+    isinvertible(t)
 
-Computes the log(abs(det(J(b(x))))) where J is the jacobian of the transform.
-Similarily for the inverse-transform.
+Return `true` if `t` is invertible, and `false` otherwise.
+"""
+isinvertible(t::Transform) = invertible(t) isa Invertible
 
-Default implementation for `Inverse{<:Bijector}` is implemented as
-`- logabsdetjac` of original `Bijector`.
 """
-logabsdetjac(ib::Inverse{<:Bijector}, y) = - logabsdetjac(ib.orig, ib(y))
+    inverse(b::Transform)
+    Inverse(b::Transform)
 
+A `Transform` representing the inverse transform of `b`.
 """
-    with_logabsdet_jacobian(b::Bijector, x)
+struct Inverse{T<:Transform} <: Transform
+    orig::T
 
-Computes both `transform` and `logabsdetjac` in one forward pass, and
-returns a named tuple `(b(x), logabsdetjac(b, x))`.
+    function Inverse(orig::Transform)
+        if !isinvertible(orig)
+            error("$(orig) is not invertible")
+        end
+
+        return new{typeof(orig)}(orig)
+    end
+end
 
-This defaults to the call above, but often one can re-use computation
-in the computation of the forward pass and the computation of the
-`logabsdetjac`. `forward` allows the user to take advantange of such
-efficiencies, if they exist.
+Functors.@functor Inverse
+
+"""
+    inverse(t::Transform[, ::Invertible])
+
+Returns the inverse of transform `t`.
 """
-with_logabsdet_jacobian(b::Bijector, x) = (b(x), logabsdetjac(b, x))
+inverse(t::Transform) = Inverse(t)
```

Comment:
Just remove this definition, and it should be sufficient to operate with `InverseFunctions.NoInverse` instead of `Invertible`/`NoInvertible`?

Reply:
Done!